### PR TITLE
20240728-EVP_PKEY_is_a

### PR DESCRIPTION
**Pull request by douzzer**
`wolfcrypt/src/evp.c`: add `EVP_PKEY_is_a()`.

### CODE REVIEW

**Comment by bandi13** [comment]
retest this please

**Comment by douzzer** [comment]
retest this please -- `PRB-aarch64-tests-v2` testsuite "wolfSSL error: tcp connect failed: Connection refused"

**Comment by douzzer** [comment]
retest this please

(unreproducible failure on `PRB-fsanitize-addr-v3` `./configure --enable-jobserver=4 CC="clang -fsanitize=address,undefined -g" --enable-all --enable-sp-math-all` `FAIL: scripts/openssl.test`)

File: wolfcrypt/src/evp.c [hunk 1]
```diff
@@ -283,6 +283,41 @@ static const struct s_ent {
 
 static const char EVP_NULL[] = "NULL";
 
+static const struct pkey_type_name_ent {
+    int type;
+    const char *name;
+} pkey_type_names[] =
+{
```
Comment (julek-wolfssl) [inline wolfcrypt/src/evp.c:290]: We usually start array definitions on the same line.

File: wolfcrypt/src/evp.c [hunk 2]
```diff
@@ -283,6 +283,41 @@ static const struct s_ent {
...
+    { EVP_PKEY_RSA,     "RSA" },
+    { EVP_PKEY_EC,      "EC" },
+    { EVP_PKEY_DH,      "DH" },
+    { EVP_PKEY_DSA,     "DSA" }
+};
+
+static int pkey_type_by_name(const char *name) {
+    unsigned int i;
+    if (name == NULL)
+        return 0;
```
Comment (julek-wolfssl) [inline wolfcrypt/src/evp.c:300]: I would return `EVP_PKEY_NONE` in the error cases in this function.

File: wolfcrypt/src/evp.c [hunk 3]
```diff
@@ -283,6 +283,41 @@ static const struct s_ent {
...
+            return pkey_type_names[i].type;
+    }
+    return 0;
+}
+
+int wolfSSL_EVP_PKEY_is_a(const WOLFSSL_EVP_PKEY *pkey, const char *name) {
+    int type;
+
+    if (pkey == NULL)
+        return 0;
```
Comment (julek-wolfssl) [inline wolfcrypt/src/evp.c:312]: I think its better to return `WOLFSSL_FAILURE` in the error cases in this function.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Final fixes and ready to merge.

File: wolfcrypt/src/evp.c (hunk 3 above)
Comment (douzzer) [reply wolfcrypt/src/evp.c]: :heavy_check_mark:

File: wolfcrypt/src/evp.c [hunk 4]
```diff
@@ -283,6 +283,40 @@ static const struct s_ent {
...
+}
+
+int wolfSSL_EVP_PKEY_is_a(const WOLFSSL_EVP_PKEY *pkey, const char *name) {
+    int type;
+
+    if (pkey == NULL)
+        return WOLFSSL_FAILURE;
+
+    type = pkey_type_by_name(name);
+    if (type == 0)
```
Comment (julek-wolfssl) [inline wolfcrypt/src/evp.c:314]:
```suggestion
    if (type == EVP_PKEY_NONE)
```

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/evp.c (hunk 4 above)
Comment (douzzer) [reply wolfcrypt/src/evp.c]: :heavy_check_mark:

**Review by julek-wolfssl — APPROVED** [review]

**Pull request merged**